actors/builtin/miner/miner_actor.go

```diff
@@ -2067,6 +2068,297 @@ func (a Actor) RepayDebt(rt Runtime, _ *abi.EmptyValue) *abi.EmptyValue {
+
+	type updateWithDetails struct {
+		update            *ReplicaUpdate
+		sectorInfo        *SectorOnChainInfo
+		dealWeight        market.SectorWeights
+		unsealedSectorCID cid.Cid
+	}
+
+	// Group declarations by deadline
+	declsByDeadline := map[uint64][]*updateWithDetails{}
+	var deadlinesToLoad []uint64
+	for i, updateWithSectorInfo := range validatedUpdates {
+		if _, ok := declsByDeadline[updateWithSectorInfo.update.Deadline]; !ok {
+			deadlinesToLoad = append(deadlinesToLoad, updateWithSectorInfo.update.Deadline)
+		}
+		declsByDeadline[updateWithSectorInfo.update.Deadline] = append(declsByDeadline[updateWithSectorInfo.update.Deadline], &updateWithDetails{
+			update:            updateWithSectorInfo.update,
+			sectorInfo:        updateWithSectorInfo.sectorInfo,
+			dealWeight:        dealWeights.Sectors[i],
+			unsealedSectorCID: unsealedSectorCIDs[i],
+		})
+	}
+
+	rewRet := requestCurrentEpochBlockReward(rt)
+	powRet := requestCurrentTotalPower(rt)
+
+	succeededSectors := bitfield.New()
+	var st State
+	rt.StateTransaction(&st, func() {
+		deadlines, err := st.LoadDeadlines(store)
+		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load deadlines")
+
+		newSectors := make([]*SectorOnChainInfo, len(validatedUpdates))
+		for _, dlIdx := range deadlinesToLoad {
+			deadline, err := deadlines.LoadDeadline(store, dlIdx)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load deadline %d", dlIdx)
+
+			partitions, err := deadline.PartitionsArray(store)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load partitions for deadline %d", dlIdx)
+
+			quant := st.QuantSpecForDeadline(dlIdx)
+
+			for _, updateWithDetails := range declsByDeadline[dlIdx] {
+				updateProofType, err := updateWithDetails.sectorInfo.SealProof.RegisteredUpdateProof()
+				builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "couldn't load update proof type")
+				builtin.RequirePredicate(rt, updateWithDetails.update.UpdateProofType == updateProofType, exitcode.ErrIllegalArgument, "unsupported update proof type %d", updateWithDetails.update.UpdateProofType)
+
+				err = rt.VerifyReplicaUpdate(
+					proof.ReplicaUpdateInfo{
+						UpdateProofType:      updateProofType,
+						NewSealedSectorCID:   updateWithDetails.update.NewSealedSectorCID,
+						OldSealedSectorCID:   updateWithDetails.sectorInfo.SealedCID,
+						NewUnsealedSectorCID: updateWithDetails.unsealedSectorCID,
+						Proof:                updateWithDetails.update.ReplicaProof,
+					})
+
+				builtin.RequireNoErr(rt, err, exitcode.ErrIllegalArgument, "failed to verify replica proof for sector %d", updateWithDetails.sectorInfo.SectorNumber)
+
+				newSectorInfo := updateWithDetails.sectorInfo
+
+				newSectorInfo.SealedCID = updateWithDetails.update.NewSealedSectorCID
+				if newSectorInfo.SectorKeyCID == nil {
+					newSectorInfo.SectorKeyCID = &updateWithDetails.sectorInfo.SealedCID
```

Comment:
updateWithDetails.sectorInfo is aliased with newSectorInfo which makes this code do the wrong thing